fleetspeak/src/e2etesting/lib/setup_components.go

```diff
@@ -0,0 +1,318 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	ptypes "github.com/golang/protobuf/ptypes"
+	duration "github.com/golang/protobuf/ptypes/duration"
+	daemonservicePb "github.com/google/fleetspeak/fleetspeak/src/client/daemonservice/proto/fleetspeak_daemonservice"
+	clientConfigPb "github.com/google/fleetspeak/fleetspeak/src/client/generic/proto/fleetspeak_client_generic"
+	"github.com/google/fleetspeak/fleetspeak/src/common"
+	spb "github.com/google/fleetspeak/fleetspeak/src/common/proto/fleetspeak"
+	cpb "github.com/google/fleetspeak/fleetspeak/src/config/proto/fleetspeak_config"
+	fcpb "github.com/google/fleetspeak/fleetspeak/src/server/components/proto/fleetspeak_components"
+	grpcServicePb "github.com/google/fleetspeak/fleetspeak/src/server/grpcservice/proto/fleetspeak_grpcservice"
+	servicesPb "github.com/google/fleetspeak/fleetspeak/src/server/proto/fleetspeak_server"
+	"google.golang.org/grpc"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"sync"
+	"time"
+)
+
+var (
+	mysqlDatabase = flag.String("mysql_database", "", "MySQL database name to use")
+	mysqlUsername = flag.String("mysql_username", "", "MySQL username to use")
+	mysqlPassword = flag.String("mysql_password", "", "MySQL password to use")
+)
+
+type componentPids struct {
+	serverPid  int
+	clientPid  int
+	servicePid int
+}
+
+func newComponentPids() componentPids {
+	return componentPids{-1, -1, -1}
+}
+
+func killProcess(pid int) {
+	process, err := os.FindProcess(pid)
+	if err != nil {
+		return
+	}
+	process.Kill()
+}
+
+func (cp *componentPids) killAll() {
+	if cp.serverPid != -1 {
+		killProcess(cp.serverPid)
+	}
+	if cp.clientPid != -1 {
+		killProcess(cp.clientPid)
+	}
+	if cp.servicePid != -1 {
+		killProcess(cp.servicePid)
+	}
+}
+
+// Starts a command and redirects its output to main stdout
+func startProcess(cmd *exec.Cmd, pidChan chan int) error {
+	stdoutIn, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	stderrIn, err := cmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	err = cmd.Start()
+	if err != nil {
+		return fmt.Errorf("cmd.Start() failed: %v", err)
+	}
+	pidChan <- cmd.Process.Pid
+	close(pidChan)
+
+	// cmd.Wait() should be called only after we finish reading
+	// from stdoutIn and stderrIn.
+	// wg ensures that we finish
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		io.Copy(os.Stdout, stdoutIn)
+		wg.Done()
+	}()
+
+	_, err = io.Copy(os.Stderr, stderrIn)
+	if err != nil {
+		return fmt.Errorf("Copy streams error: %v", err)
+	}
+
+	wg.Wait()
+	cmd.Wait()
+	return nil
+}
+
+func getNewClientID(admin servicesPb.AdminClient, startTime time.Time) (string, error) {
```

Comment:
Please return nil instead of "" in case an error is returned.

(All occurences in this PR)

Reply:
As I know, "" is the default value for string and nil cannot be assigned to strings - string is not a pointer type. Or do you mean that it's better to return a pointer to string? Why?

Reply:
Alexander, and the code, is correct on this point. The go standard is simply that the returned value is meaningless and should not be used on error, unless clearly documented otherwise.

FS preferes to return default values - nil for refs, or "" for strings, 0 for ints, etc, on error. This is mostly to make it more likely that forgetting to handle an error would be noticed.